File: javascript/src/parallelization.ts
import { PromptTemplate } from "@langchain/core/prompts";
import { ChatOpenAI } from "@langchain/openai";
import { StringOutputParser } from "@langchain/core/output_parsers";
import { Document } from "@langchain/core/documents";
import { CheerioWebBaseLoader } from "@langchain/community/document_loaders/web/cheerio";
import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";
import dotenv from "dotenv";
import path from "path";
import { traceable } from "langsmith/traceable";

dotenv.config({ path: path.resolve(__dirname, "../../.env") });

const model = new ChatOpenAI({ model: "gpt-4o-mini", temperature: 0 });
const parser = new StringOutputParser();

async function summarize_chunk(document: Document) {
  const prompt = PromptTemplate.fromTemplate(
    "Write a concise summary of the following: {chunk}"
  );
  const chain = prompt.pipe(model).pipe(parser);
  return await chain.invoke({ chunk: document.pageContent });
}

async function aggregate_summaries(summaries: string[]) {
  const prompt = PromptTemplate.fromTemplate(
    `The following is a set of summaries:
{docs}
Take these and distill it into a final, consolidated summary of the main themes.`
  );
  const chain = prompt.pipe(model).pipe(parser);
  return await chain.invoke({ docs: summaries.join("\n\n") });
}

async function parallelization(url: string) {
  async function run(input: { url: string }) {
    const loader = new CheerioWebBaseLoader(input.url);
    const docs = await loader.load();
    const text_splitter = new RecursiveCharacterTextSplitter({
      chunkSize: 5000,
      chunkOverlap: 200,
    });
    const texts = await text_splitter.splitDocuments(docs);
    const summaries = await Promise.all(texts.map(summarize_chunk));
    return await aggregate_summaries(summaries);
  }
  return traceable(run, { name: "parallelization" })({ url });
}

async function main() {
  const url = "https://lilianweng.github.io/posts/2023-06-23-agent/";
  const summary = await parallelization(url);
  console.log(summary);
}

main();

File: python/langchain/parallelization.py
# https://www.agentrecipes.com/parallelization
# Parallelization takes advantage of tasks that can broken up into discrete independent parts.
# The user's prompt is passed to multiple LLMs simultaneously. Once all the LLMs respond,
# their answers are all sent to a final LLM call to be aggregated for the final answer.

import asyncio
from langchain_openai import ChatOpenAI
from langchain_core.output_parsers import StrOutputParser
from langsmith import traceable
from langchain_community.document_loaders import WebBaseLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_core.prompts import PromptTemplate
from langchain_core.documents import Document
from typing import List
from dotenv import load_dotenv

load_dotenv()

# In this example, we are analyzing a lengthy document by dividing
# it into sections and assigning each section to a separate LLM for summarization,
# then combining the summaries into a comprehensive overview.

model = ChatOpenAI(model="gpt-4o-mini", temperature=0)
parser = StrOutputParser()


@traceable(name="summarize_chunk")
async def summarize_chunk(document: Document) -> str:
    prompt = PromptTemplate.from_template(
        "Write a concise summary of the following: {chunk}"
    )
    chain = prompt | model | parser
    return await chain.ainvoke({"chunk": document.page_content})


@traceable(name="aggregate_summaries")
async def aggregate_summaries(summaries: List[str]) -> str:
    prompt = PromptTemplate.from_template(
        """
The following is a set of summaries:
{docs}
Take these and distill it into a final, consolidated summary
of the main themes.
"""
    )
    chain = prompt | model | parser
    return await chain.ainvoke({"docs": "\n\n".join(summaries)})


@traceable(name="parallelization")
async def parallelization(url: str) -> str:
    """Create a workflow that uses multiple LLMs to analyze a URL and return a summary of the content."""
    loader = WebBaseLoader(url)
    docs = loader.load()
    text_splitter = RecursiveCharacterTextSplitter(chunk_size=5000, chunk_overlap=200)
    texts = text_splitter.split_documents(docs)
    # Gather and await all summary tasks
    summaries = await asyncio.gather(*[summarize_chunk(text) for text in texts])
    return await aggregate_summaries(summaries)


async def main():
    url = "https://www.anthropic.com/research/building-effective-agents"
    summary = await parallelization(url)
    print(summary)


if __name__ == "__main__":
    asyncio.run(main())
